src/main/java/net/neoforged/neoforge/flag/FlagManager.java

```diff
@@ -0,0 +1,225 @@
+     *
+     * @param flag {@link Flag} which to have its state toggled.
+     * @return {@code true} if the state of the {@link Flag} changed, {@code false} otherwise.
+     */
+    public boolean toggle(Flag flag) {
+        return set(flag, !flags.getBoolean(flag));
+    }
+
+    /**
+     * Returns an immutable collection containing all currently enabled {@link Flag Flags}.
+     *
+     * @return Immutable collection containing all currently enabled {@link Flag Flags}.
+     */
+    public Collection<Flag> getEnabledFlags() {
+        return enabledFlags().toList();
+    }
+
+    /**
+     * Returns {@link Stream} representing all currently enabled {@link Flag Flags}.
+     *
+     * @return {@link Stream} representing all currently enabled {@link Flag Flags}.
+     */
+    public Stream<Flag> enabledFlags() {
+        return flagsView.stream().filter(this::isEnabled);
+    }
+
+    /**
+     * Loads all relevant data from the given level via the {@link NeoForgeMod#LEVEL_FLAGS flags} data attachment.
+     *
+     * @param level The level in which to load data from.
+     */
+    @ApiStatus.Internal
+    public void loadFromLevel(Level level) {
+        level.getExistingData(NeoForgeMod.LEVEL_FLAGS).ifPresent(enabledFlags -> {
+            reset();
+            enabledFlags.forEach(flag -> flags.put(flag, true));
+        });
+    }
+
+    /**
+     * Saves all relevant data to the given level via the {@link NeoForgeMod#LEVEL_FLAGS flags} data attachment.
+     *
+     * @param level The level in which to save data to.
+     */
+    @ApiStatus.Internal
+    public void saveToLevel(Level level) {
+        level.setData(NeoForgeMod.LEVEL_FLAGS, enabledFlags().collect(Collectors.toSet()));
+    }
+
+    void syncFromRemote(ClientboundSyncFlag payload) {
+        flags.put(payload.flag, payload.enabled);
+    }
+
+    void syncFromRemote(ClientboundSyncFlags payload) {
+        reset();
+        payload.enabledFlags.forEach(flag -> flags.put(flag, true));
+    }
+
+    /**
+     * Clears all loaded flag states for this manager.
+     *
+     * @apiNote This will only clear for the current {@link Dist phsyical side}.
+     */
+    @ApiStatus.Internal
+    public void reset() {
+        flags.clear();
+    }
+
+    /**
+     * Prefer to lookup instance via {@linkplain ILevelReaderExtension#getModdedFlagManager()}
+     */
+    @UnstableApi
+    public static Optional<FlagManager> lookup() {
+        return lookupClient().or(() -> {
+            var server = ServerLifecycleHooks.getCurrentServer();
+            return server == null ? Optional.empty() : Optional.of(server.getModdedFlagManager());
+        });
+    }
+
+    private static Optional<FlagManager> lookupClient() {
+        if (DatagenModLoader.isRunningDataGen())
+            return Optional.empty();
+        if (FMLEnvironment.dist.isClient())
+            return Optional.ofNullable(Minecraft.getInstance()).map(Minecraft::getModdedFlagManager);
```

Comment:
I'm not sure the way this is implemented is safe on a dedicated server despite being behind a dist check. Have you tested this on a dedicated server?

Reply:
I have tested quite a few times in a ded server or more specifically the `Tests: Server` run, and seems to run fine with no dist issues.